suitcase/tiff/__init__.py

```diff
@@ -1,25 +1,256 @@
+    """
+    Serialize a stream of documents to TIFF stack(s).
+
+    This creates a file named:
+    ``<directory>/<file_prefix>{stream_name}-{field}.tiff``
+    for every Event stream and field that contains 2D 'image like' data.
+
+    .. warning::
+
+        This process explicitly ignores all data that is not 2D and does not
+        include any metadata in the output file.
+
+
+    .. note::
+
+        This can alternatively be used to write data to generic buffers rather
+        than creating files on disk. See the documentation for the
+        ``directory`` parameter below.
+
+    Parameters
+    ----------
+    gen : generator
+        xpected to yield ``(name, document)`` pairs
+
+    directory : string, Path or Manager.
+        For basic uses, this should be the path to the output directory given
+        as a string or Path object. Use an empty string ``''`` to place files
+        in the current working directory.
+
+        In advanced applications, this may direct the serialized output to a
+        memory buffer, network socket, or other writable buffer. It should be
+        an instance of ``suitcase.utils.MemoryBufferManager`` and
+        ``suitcase.utils.MultiFileManager`` or any object implementing that
+        inferface. See the suitcase documentation (LINK ONCE WRITTEN) for
+        details.
+
+    file_prefix : str, optional
+        The first part of the filename of the generated output files. This
+        string may include templates as in ``{proposal_id}-{sample_name}-``,
+        which are populated from the RunStart document. The default value is
+        ``{uid}-`` which is guaranteed to be present and unique. A more
+        descriptive value depends on the application and is therefore left to
+        the user.
+
+    **kwargs : kwargs
+        kwargs to be passed to ``tifffile.TiffWriter.save``.
+    """
+    def __init__(self, directory, file_prefix='{uid}-', **kwargs):
+
+        if isinstance(directory, (str, Path)):
+            self.manager = suitcase.utils.MultiFileManager(directory)
+        else:
+            self.manager = directory
+
+        self.artifacts = self.manager._artifacts
+        self._stream_names = {}  # maps stream_names to each descriptor uids
+        self._files = defaultdict(dict)  # maps stream_name to field/file dicts
+        self._filenames = {}  # map stream_name to file names of tiff files
+        self._file_prefix = file_prefix
+        self._templated_file_prefix = ''
+        self._kwargs = kwargs
+        self._start_found = False
+
+    def start(self, doc):
+        '''Extracts `start` document information for formatting file_prefix.
+
+        This method checks that only one `start` document is seen and formats
+        `file_prefix` based on the contents of the `start` document.
+
+        Parameters:
+        -----------
+        doc : dict
+            RunStart document
+        '''
+
+        # raise an error if this is the second `start` document seen.
```

Comment:
I think we still need some thought about how exactly pushing multiple runs through will work.

Reply:
agreed, for now I think not allowing it, in order to avoid overwriting files, is the best option  but in the future we should discuss how to best allow it.

Reply:
The issue isn't with overwriting files: recall that `file_prefix='{uid}-'` addresses that. The issue is an every expanding cache of held resources. For example:

```py
# Automatically export all files in the background.
# (Of course in practice we would do this across a 0MQ socket, but the
# principle is the same.)
serializer = Serizlier(...)
RE.subscribe(serializer)
```

The `serializer` instance will hold onto an ever-increasing set of `artifacts`.

Reply:
Can we address this using `RunRouter` to create a new `Serializer` for every run?

Reply:
Yes, but that would mean that we'd need to use it inside `export` above, which would mean we'd want to move it into `event-model` to avoid a bluesky dep.

Reply:
N.B. @awalter-bnl that as written `export(db[-1], '')` works but `export(db[-2:], '')` will explode. Seems like we want to fix that before declaring victory here.

Reply:
Or maybe this would be fine for now?

```
for h in db[-2:]:
    export(h, '')
```

Reply:
either solution is fine with me, what is the consensus?

Reply:
For the record, the consensus between us was: Support "one run only" for now; likely add support for multiple runs later after the `RunRouter` development has solidified.